### ISSUE TITLE
Enhancement: please make the urls seo friendly with pretty urls

### ISSUE THREAD

**Issue by harshitjawla** [humhub/humhub-modules-wiki#136]
Hello. Please fix my problem.
I eagerly want the feature in which we can see user posts with their own separate title. For. Ex. When a user posts in wiki, in title bar of the browser post title should be seen instead of just space name.
Also please remove that queries (?) From urls... And make the title just like this:
https://demo.humhub.com/s/space-name/this-is-topic-title
It would help me a lot much..please try to do this..  very soon
 thanks in advance.

**Comment by harshitjawla** [comment]
@buddh4 @luke- Please help by enhanching..this module...v. much. Thanks for your reply

**Comment by luke-** [comment]
I think this would be a good URL for wiki pages:
https://demo.humhub.com/s/space-name/wiki/some-page-name

I remember that there was a reason not to implement this way directly at that time...

 But we should try it again. Of course, it is important that all special characters are handled correctly...

**Comment by yurabakhtin** [comment]
@luke- What I have investigated:

- Currently we cannot parse urls like `/s/space-name/wiki/some-page-name` from module side, because such urls is parsed by Space URL Manager here https://github.com/humhub/humhub/blob/master/protected/humhub/modules/space/components/UrlRule.php#L63, same for User Profile urls started with `/u/`, I.e. if URL is started with `/s/` or `/u/` then it cannot be parsed to other modules.

I see 2 solutions:

1. Instead of the URL `/s/space-name/wiki/some-page-name` we might use `/wiki/some-page-name`. This way already works on the module "Custom Pages" for URLs like `/p/123` - https://github.com/humhub/custom-pages/blob/master/components/PageUrlRule.php#L62.
2. We should implement a new event trigger in order to handle the result `return [$parts[2], $params];` here https://github.com/humhub/humhub/blob/master/protected/humhub/modules/space/components/UrlRule.php#L75 (and on User Profile Manager Url as well). So after this modification in core files of the containers I think modules will can parse own URLs under Space or Profile urls.

It will works in both cases, but if wiki pages are linked to a container so it would be better to keep container data in URLs as well. Do you agree to implement such new core trigger?

---

Please note above we tell about URLs like `/wiki/some-page-name` but really Wiki Page has no such url data `some-page-name` in DB, it has only title, and title field is used to find a wiki page in DB, so url will be like `/wiki/Some+page+name` (`+` is encoded space char).
Yes, we can try to generate the url part `some-page-name` from wiki page title `Some page name`, but then we should think how to convert a char `-` if it is detected in wiki page title, probably we can use double `--`, any idea?

Also I see the DB column `wiki_page.title` is not unique column, however the edit form doesn't allow to save a wiki page with same title, that's probably ok, but it would be better to keep the column unique, either we should implement some new column `wiki.url` like we have for Space table column `space.url`.

**Comment by luke-** [comment]
@yurabakhtin Thank you for the Investigation. I didn't realize that this is a general problem with URL rules for ContentContainer.
I would suggest we put this issue `onHold` for now and try to create a way in the core repository to set such custom URL rules.

However, I'm not sure if an event trigger is a convenient solution here. I'll create an issue and we can discuss ideas there.

**Comment by lisandi** [comment]
What options exist?

if you would continue using one and two letter URLS you could achieve both!

u - user
p - page
w - wiki
wp - wikipage (if really necessary - why can't a subpage not be also a "w")
g - group
d - discussion
s - space
...
wp - wiki page

Having those at first place always would make clear what kind of URL / Page it is

Next the space or profile name could follow and the title.

so instead of having

[domain.tld]/s/[spacename]/w/[title]
you would have a much shorter and easier to remember
[domain.tld]/w/[spacename]-[title] or [domain.tld]/w/[spacename]/[title]

Besides that something like already existing in any Wordpress site where the URL gets simply checked and if not existing the nearest match gets suggested and shown would also help for SEO.

**Comment by luke-** [comment]
@lisandi Yes, this issue is about shortening and improving URLs within a space/user. See more here: https://github.com/humhub/wiki/issues/136#issuecomment-882147326

**Comment by luke-** [comment]
@lisandi  Hmm,  according to the current URL concept, Space URLs should always start with https://example.com/s/spacename and user profile URLs should always start with https://example.com/u/username.

This way you can see directly by the URL in which context you are.
https://example.com/s/spacename/w/pagetitle

**Comment by lisandi** [comment]
@luke- Ok I understand - nevertheless the URLs are all unnecessary long in humhub and ways to shorten that would be more than appreciated. you really don't meed long words like "wiki" it is like the "www' with URLs - keeping it short and simple with a clear structure would be nice.

If the structure is s/[spacename]/w/[wikititle] it already would be way better than right now.
The same applies to u/[username]/w/[wikititle] and i.e. /b/ for blog /p/ for custom pages etc

than inside a space or user which seem to be the two main base groups you would have i.e. global items like /d/ discussions or /f/ forums and specific for users or spaces they could appear in the same way behind the user or space name.

I just was setting links to permalink and here seems to be the same problem as /content/perma?id=50 is way to long and /cp50 would be much better or cp/50 

I tried to test the suggested wiki improvement with URLs but got an error message as one file seems to be missing to get that working. It might appears in 1.9 but could not find it in one of the repos.

`Error: Interface 'humhub\components\ContentContainerUrlRuleInterface' not found in /[path]/htdocs/protected/modules/wiki/components/WikiPageUrlRule.php:23`

**Comment by luke-** [comment]
@lisandi The URL improvements will first be available in v1.10.
